src/v/kafka/server/group.cc

```diff
@@ -1206,6 +1221,73 @@ group::begin_tx(cluster::begin_group_tx_request r) {
     co_return reply;
 }
 
+ss::future<cluster::prepare_group_tx_reply>
+group::prepare_tx(cluster::prepare_group_tx_request r) {
+    if (_partition->term() != _term) {
+        co_return make_prepare_tx_reply(cluster::tx_errc::timeout);
+    }
+
+    if (r.etag != _term) {
+        co_return make_prepare_tx_reply(cluster::tx_errc::timeout);
+    }
+
+    auto tx_it = _volatile_txs.find(r.pid);
+    if (tx_it == _volatile_txs.end()) {
+        co_return make_prepare_tx_reply(cluster::tx_errc::timeout);
+    }
+
+    auto tx_entry = group_log_prepared_tx{.group_id = r.group_id, .pid = r.pid};
+
+    for (const auto& [tp, offset] : tx_it->second.offsets) {
+        group_log_prepared_tx_offset tx_offset;
+
+        tx_offset.tp = tp;
+        tx_offset.offset = offset.offset;
+        tx_offset.leader_epoch = offset.leader_epoch;
+        tx_offset.metadata = offset.metadata;
+        tx_entry.offsets.push_back(tx_offset);
+    }
+
+    volatile_tx tx = tx_it->second;
+    _volatile_txs.erase(tx_it);
+
+    iobuf key;
+    kafka::response_writer w(key);
+    // TODO: https://app.clubhouse.io/vectorized/story/2200
+    // include producer_id+type into key to make it unique-ish
+    // to prevent being GCed by the compaction
+    w.write(prepared_tx_record_version());
```

Comment:
even abort/commit are internal for consumer groups, I was planning to take care of this in the linked clubhouse issue

Reply:
I see. if they are internal we should probably use our internal serialization protocol. iirc abort/commit are returned to clients, so that would be the exception.

Reply:
will change it in a couple PR when take care after the linked clubhouse issue

afaik we return abort/commit markers only when we read from data partition and consumer group's topic is hidden behind the consumer group api

Reply:
its partially true. (1) there are 3rd party tools that read the consumer group topic directly and (2) we know that abort/commit markers will be in normal topics and sent back to clients so no reason to have two different encoding for the same fundamental batch type?

Reply:
Right now I use custom format for the abort markers to pass information necessary for processing Redpanda transaction. It's possible to reconstruct this information from previous tx messages so we may use the Kafka's abort marker format. But even then there will be issues with clients reading from Kafka's internal topic and expecting a certain format. For example, committed offsets in Redpanda live in the prepared batches which are opaque to Kafka clients so the clients will be blind to any committed offsets.

Reply:
we'll probably end up migrating to a format that is compatible with external tools. but in that case I think we can continue to have internal-only data in control batches that are filtered out.

Reply:
Good point, we may use compatible format of do some transformations like we do with the offsets, let's iterate on this later https://github.com/vectorizedio/redpanda/issues/1286 so far this pr doesn't make __consumer_offsets less compatible if the clients don't use transactions